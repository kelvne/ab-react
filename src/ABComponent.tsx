import * as React from 'react';
import { useMemo } from 'react';

import { ABComponentProps } from './types';
import ChildrenHelper from './ChildrenHelper';
import WithVersionDefined from './WithVersionDefined';
import WithoutVersionDefined from './WithoutVersionDefined';

/**
 * ABComponent is the component that decides if the stateful or stateless version
 * of the element will be rendered.
 */
function ABComponent({
  cookieName, children, event, forceVersion,
}: ABComponentProps) {
  const helper = useMemo(() => new ChildrenHelper(children), [children]);
  helper.warnMoreThanTwo();

  if (typeof forceVersion === 'number' && forceVersion >= 0) {
    return (
      <WithVersionDefined
        childrenHelper={helper}
        forceVersion={forceVersion}
        event={event}
      />
    );
  }

  return (
    <WithoutVersionDefined
      childrenHelper={helper}
      cookieName={cookieName}
      event={event}
    />
  );
}

export default ABComponent;
