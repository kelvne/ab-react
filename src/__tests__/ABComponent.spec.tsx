import * as React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import ABComponent from '../ABComponent';

jest.mock('../WithoutVersionDefined', () => function WVD() { return (<div>WithoutVersionDefined</div>); });

describe('ABComponent', () => {
  const cookieName = 'cookie-1';
  const children = [<div key="k-1">a</div>, <div key="k-2">b</div>];

  test('expect to render WithVersionDefined if forceVersion is set', () => {
    const { container } = render(
      <ABComponent
        cookieName={cookieName}
        forceVersion={0}
      >
        {children.map((el) => el)}
      </ABComponent>,
    );

    expect(container).toMatchSnapshot();
  });

  test('expect to render WithoutVersionDefined if forceVersion is not set', () => {
    const { getByText } = render(
      <ABComponent cookieName={cookieName}>
        {children.map((el) => el)}
      </ABComponent>,
    );

    expect(getByText('WithoutVersionDefined')).toBeInTheDocument();
  });
});
