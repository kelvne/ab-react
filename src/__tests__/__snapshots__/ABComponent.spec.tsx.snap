// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ABComponent expect to render WithVersionDefined if forceVersion is set 1`] = `
<div>
  <div>
    a
  </div>
</div>
`;
